javalib/src/main/scala/java/math/BigInteger.scala

```diff
@@ -0,0 +1,1500 @@
+   * big to be represented as an float, then {@code Float.POSITIVE_INFINITY} or
+   * {@code Float.NEGATIVE_INFINITY} is returned. Note, that not all integers x
+   * in the range [-Float.MAX_VALUE, Float.MAX_VALUE] can be represented as a
+   * float. The float representation has a mantissa of length 24. For example,
+   * 2^24+1 = 16777217 is returned as float 16777216.0.
+   *
+   * @return this {@code BigInteger} as a float value.
+   */
+  override def floatValue(): Float = {
+    java.lang.Float.parseFloat(this.toString)
+  }
+
+  /**
+   * Returns a new {@code BigInteger} whose value is greatest common divisor of
+   * {@code this} and {@code val}. If {@code this==0} and {@code val==0} then
+   * zero is returned, otherwise the result is positive.
+   *
+   * @param bi value with which the greatest common divisor is computed.
+   * @return {@code gcd(this, val)}.
+   * @throws NullPointerException if {@code val == null}.
+   */
+  def gcd(bi: BigInteger): BigInteger = {
+    val val1 = this.abs()
+    val val2 = bi.abs()
+    // To avoid a possible division by zero
+    if (val1.signum() == 0) {
+      val2
+    } else if (val2.signum() == 0) {
+      val1
+    } else if (((val1.numberLength == 1) || ((val1.numberLength == 2) && (val1.digits(1) > 0))) &&
+        (val2.numberLength == 1 || (val2.numberLength == 2 && val2.digits(1) > 0))) {
+      // Optimization for small operands
+      // (op2.bitLength() < 64) and (op1.bitLength() < 64)
+      BigInteger.valueOf(Division.gcdBinary(val1.longValue(), val2.longValue()))
+    } else {
+      Division.gcdBinary(val1.copy(), val2.copy())
+    }
+  }
+
+  /**
+   * Returns the position of the lowest set bit in the two's complement
+   * representation of this {@code BigInteger}. If all bits are zero (this=0)
+   * then -1 is returned as result.
+   * <p>
+   * <b>Implementation Note:</b> Usage of this method is not recommended as the
+   * current implementation is not efficient.
+   *
+   * @return position of lowest bit if {@code this != 0}, {@code -1} otherwise
+   */
+  def getLowestSetBit(): Int = {
+    if (sign == 0) {
+      -1
+    } else {
+      // (sign != 0) implies that exists some non zero digit
+      val i = getFirstNonzeroDigit
+      (i << 5) + java.lang.Integer.numberOfTrailingZeros(digits(i))
+    }
+  }
+
+  /**
+   * Returns a hash code for this {@code BigInteger}.
+   *
+   * @return hash code for {@code this}.
+   */
+  override def hashCode(): Int = {
+    if (_hashCode != 0) {
+      _hashCode
+    } else {
+      for (i <- 0 until digits.length) {
+        _hashCode = _hashCode * 33 + (digits(i) & 0xffffffff)
+      }
+      _hashCode = _hashCode * sign
+      _hashCode
+    }
+  }
+
+  /**
+   * Returns this {@code BigInteger} as an int value. If {@code this} is too big
+   * to be represented as an int, then {@code this} % 2^32 is returned.
+   *
+   * @return this {@code BigInteger} as an int value.
+   */
+  override def intValue(): Int = sign * digits(0)
```

Comment:
Does this work for values bigger than `Int.MaxValue` whose 31th bit is 1? For example, 0x80000000L in BigInteger?